// @backstage/plugin-kubernetes-react/dist/components/PodExecTerminal/PodExecTerminalDialog.esm.js
import { jsx } from 'react/jsx-runtime';
import OpenInBrowserIcon from '@material-ui/icons/OpenInBrowser';
import { KubernetesDialog } from '../KubernetesDialog/KubernetesDialog.esm.js';
import '@backstage/core-plugin-api';
import { useIsPodExecTerminalSupported } from '../../hooks/useIsPodExecTerminalSupported.esm.js';
import 'react';
import 'react-use/esm/useInterval';
import 'react-use/esm/useAsyncRetry';
import '../../kubernetes-auth-provider/types.esm.js';
import '../../api/types.esm.js';
import '../../hooks/PodNamesWithErrors.esm.js';
import '../../hooks/PodNamesWithMetrics.esm.js';
import '../../hooks/GroupedResponses.esm.js';
import '../../hooks/Cluster.esm.js';
import '../../hooks/usePodMetrics.esm.js';
import '../../hooks/useMatchingErrors.esm.js';
import { PodExecTerminal } from './PodExecTerminal.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { kubernetesReactTranslationRef } from '../../translation.esm.js';

const PodExecTerminalDialog = (props) => {
  const { cluster, containerName, podName } = props;
  const { t } = useTranslationRef(kubernetesReactTranslationRef);
  const isPodExecTerminalSupported = useIsPodExecTerminalSupported();
  return !isPodExecTerminalSupported.loading && isPodExecTerminalSupported.value && /* @__PURE__ */ jsx(
    KubernetesDialog,
    {
      buttonAriaLabel: t("podExecTerminal.buttonAriaLabel"),
      buttonIcon: /* @__PURE__ */ jsx(OpenInBrowserIcon, {}),
      buttonText: t("podExecTerminal.buttonText"),
      disabled: isPodExecTerminalSupported.loading || !isPodExecTerminalSupported.value,
      title: t("podExecTerminal.titleTemplate", {
        podName,
        containerName,
        clusterName: cluster.title || cluster.name
      }),
      children: /* @__PURE__ */ jsx(PodExecTerminal, { ...props })
    }
  );
};

export { PodExecTerminalDialog };
