// @backstage/plugin-kubernetes-react/dist/components/Pods/PodsTable.esm.js
import { jsx, Fragment } from 'react/jsx-runtime';
import { useContext } from 'react';
import { PodDrawer } from './PodDrawer/PodDrawer.esm.js';
import { Table } from '@backstage/core-components';
import '@material-ui/core/Card';
import '@material-ui/core/CardActions';
import '@material-ui/core/CardContent';
import '@material-ui/core/CardHeader';
import '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import 'luxon';
import '@backstage/core-plugin-api';
import 'react-use/esm/useAsync';
import '../../api/types.esm.js';
import 'react-use/esm/useInterval';
import 'react-use/esm/useAsyncRetry';
import '../../kubernetes-auth-provider/types.esm.js';
import '../../hooks/PodNamesWithErrors.esm.js';
import '../../hooks/PodNamesWithMetrics.esm.js';
import '../../hooks/GroupedResponses.esm.js';
import { ClusterContext } from '../../hooks/Cluster.esm.js';
import { usePodMetrics } from '../../hooks/usePodMetrics.esm.js';
import { useMatchingErrors } from '../../hooks/useMatchingErrors.esm.js';
import '@material-ui/icons/OpenInBrowser';
import '../KubernetesDialog/KubernetesDialog.esm.js';
import '../PodExecTerminal/PodExecTerminal.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { kubernetesReactTranslationRef } from '../../translation.esm.js';
import '@material-ui/core/Paper';
import '@material-ui/lab/Skeleton';
import '@material-ui/icons/Subject';
import '@material-ui/core/List';
import '@material-ui/core/ListItem';
import { containersReady, totalRestarts, containerStatuses, podStatusToCpuUtil, podStatusToMemoryUtil } from '../../utils/pod.esm.js';

const READY_COLUMNS = "READY";
const RESOURCE_COLUMNS = "RESOURCE";
const PodDrawerTrigger = ({ pod }) => {
  const errors = useMatchingErrors({
    kind: "Pod",
    apiVersion: "v1",
    metadata: pod.metadata
  });
  return /* @__PURE__ */ jsx(
    PodDrawer,
    {
      podAndErrors: {
        pod,
        cluster: useContext(ClusterContext),
        errors
      }
    }
  );
};
const Cpu = ({ clusterName, pod }) => {
  const metrics = usePodMetrics(clusterName, pod);
  const { t } = useTranslationRef(kubernetesReactTranslationRef);
  if (!metrics) {
    return /* @__PURE__ */ jsx(Typography, { children: t("podsTable.unknown") });
  }
  return /* @__PURE__ */ jsx(Fragment, { children: podStatusToCpuUtil(metrics) });
};
const Memory = ({ clusterName, pod }) => {
  const metrics = usePodMetrics(clusterName, pod);
  const { t } = useTranslationRef(kubernetesReactTranslationRef);
  if (!metrics) {
    return /* @__PURE__ */ jsx(Typography, { children: t("podsTable.unknown") });
  }
  return /* @__PURE__ */ jsx(Fragment, { children: podStatusToMemoryUtil(metrics) });
};
const PodsTable = ({ pods, extraColumns = [] }) => {
  const cluster = useContext(ClusterContext);
  const { t } = useTranslationRef(kubernetesReactTranslationRef);
  const READY = [
    {
      title: t("podsTable.columns.containersReady"),
      align: "center",
      render: containersReady,
      width: "auto"
    },
    {
      title: t("podsTable.columns.totalRestarts"),
      align: "center",
      render: totalRestarts,
      type: "numeric",
      width: "auto"
    }
  ];
  const defaultColumns = [
    {
      title: t("podsTable.columns.id"),
      field: "metadata.uid",
      hidden: true
    },
    {
      title: t("podsTable.columns.name"),
      highlight: true,
      render: (pod) => {
        return /* @__PURE__ */ jsx(PodDrawerTrigger, { pod });
      }
    },
    {
      title: t("podsTable.columns.phase"),
      render: (pod) => pod.status?.phase ?? t("podsTable.unknown"),
      width: "auto"
    },
    {
      title: t("podsTable.columns.status"),
      render: containerStatuses
    }
  ];
  const columns = [...defaultColumns];
  if (extraColumns.includes(READY_COLUMNS)) {
    columns.push(...READY);
  }
  if (extraColumns.includes(RESOURCE_COLUMNS)) {
    const resourceColumns = [
      {
        title: t("podsTable.columns.cpuUsage"),
        render: (pod) => {
          return /* @__PURE__ */ jsx(Cpu, { clusterName: cluster.name, pod });
        },
        width: "auto"
      },
      {
        title: t("podsTable.columns.memoryUsage"),
        render: (pod) => {
          return /* @__PURE__ */ jsx(Memory, { clusterName: cluster.name, pod });
        },
        width: "auto"
      }
    ];
    columns.push(...resourceColumns);
  }
  const tableStyle = {
    minWidth: "0",
    width: "100%"
  };
  return /* @__PURE__ */ jsx("div", { style: tableStyle, children: /* @__PURE__ */ jsx(
    Table,
    {
      options: { paging: true, search: false, emptyRowsWhenPaging: false },
      data: pods.map((pod) => ({
        ...pod,
        id: pod?.metadata?.uid
      })),
      columns
    }
  ) });
};

export { PodsTable, READY_COLUMNS, RESOURCE_COLUMNS };
