// @backstage/plugin-kubernetes-react/dist/components/Pods/PodLogs/PodLogsDialog.esm.js
import { jsx } from 'react/jsx-runtime';
import SubjectIcon from '@material-ui/icons/Subject';
import { KubernetesDialog } from '../../KubernetesDialog/KubernetesDialog.esm.js';
import { PodLogs } from './PodLogs.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { kubernetesReactTranslationRef } from '../../../translation.esm.js';

const PodLogsDialog = ({ containerScope }) => {
  const { t } = useTranslationRef(kubernetesReactTranslationRef);
  return /* @__PURE__ */ jsx(
    KubernetesDialog,
    {
      buttonAriaLabel: t("podLogs.buttonAriaLabel"),
      buttonIcon: /* @__PURE__ */ jsx(SubjectIcon, {}),
      buttonText: t("podLogs.buttonText"),
      disabled: false,
      title: t("podLogs.titleTemplate", {
        podName: containerScope.podName,
        containerName: containerScope.containerName,
        clusterName: containerScope.cluster.title || containerScope.cluster.name
      }),
      children: /* @__PURE__ */ jsx(PodLogs, { containerScope })
    }
  );
};

export { PodLogsDialog };
