// @backstage/plugin-kubernetes-react/dist/components/Pods/PodLogs/PodLogs.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { DismissableBanner, EmptyState, LogViewer } from '@backstage/core-components';
import Paper from '@material-ui/core/Paper';
import Skeleton from '@material-ui/lab/Skeleton';
import { usePodLogs } from './usePodLogs.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { kubernetesReactTranslationRef } from '../../../translation.esm.js';

const PodLogs = ({
  containerScope,
  previous
}) => {
  const { t } = useTranslationRef(kubernetesReactTranslationRef);
  const { value, error, loading } = usePodLogs({
    containerScope,
    previous
  });
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    error && /* @__PURE__ */ jsx(
      DismissableBanner,
      {
        ...{
          message: error.message,
          variant: "error",
          fixed: false
        },
        id: "pod-logs"
      }
    ),
    /* @__PURE__ */ jsxs(
      Paper,
      {
        elevation: 1,
        style: { height: "100%", width: "100%", minHeight: "55rem" },
        children: [
          loading && /* @__PURE__ */ jsx(Skeleton, { variant: "rect", width: "100%", height: "100%" }),
          !loading && value !== void 0 && (value.text === "" ? /* @__PURE__ */ jsx(
            EmptyState,
            {
              missing: "data",
              title: t("podLogs.title"),
              description: t("podLogs.description")
            }
          ) : /* @__PURE__ */ jsx(LogViewer, { text: value.text }))
        ]
      }
    )
  ] });
};

export { PodLogs };
