// @backstage/plugin-notifications/dist/alpha.esm.js
import { jsx } from 'react/jsx-runtime';
import { PageBlueprint, ApiBlueprint, fetchApiRef, discoveryApiRef, createFrontendPlugin } from '@backstage/frontend-plugin-api';
import { rootRouteRef } from './routes.esm.js';
import { notificationsApiRef } from './api/NotificationsApi.esm.js';
import { NotificationsClient } from './api/NotificationsClient.esm.js';
export { notificationsTranslationRef } from './translation.esm.js';

const page = PageBlueprint.make({
  params: {
    path: "/notifications",
    routeRef: rootRouteRef,
    loader: () => import('./components/NotificationsPage/index.esm.js').then((m) => /* @__PURE__ */ jsx(m.NotificationsPage, {}))
  }
});
const api = ApiBlueprint.make({
  params: (defineParams) => defineParams({
    api: notificationsApiRef,
    deps: { discoveryApi: discoveryApiRef, fetchApi: fetchApiRef },
    factory: ({ discoveryApi, fetchApi }) => new NotificationsClient({ discoveryApi, fetchApi })
  })
});
var alpha = createFrontendPlugin({
  pluginId: "notifications",
  info: { packageJson: () => import('./package.json.esm.js') },
  routes: {
    root: rootRouteRef
  },
  // TODO(Rugvip): Nav item (i.e. NotificationsSidebarItem) currently needs to be installed manually
  extensions: [page, api]
});

export { alpha as default };
