// @backstage/plugin-catalog-backend/dist/actions/createRegisterCatalogEntitiesAction.cjs.js
'use strict';

var errors = require('@backstage/errors');

const isValidUrl = (url) => {
  try {
    new URL(url);
    return true;
  } catch (error) {
    return false;
  }
};
const createRegisterCatalogEntitiesAction = ({
  catalog,
  actionsRegistry
}) => {
  actionsRegistry.register({
    name: "register-entity",
    title: "Register entity in the Catalog",
    attributes: {
      destructive: false,
      readOnly: false,
      idempotent: false
    },
    description: `Registers one or more entities in the Backstage catalog by creating a Location entity that points to a remote catalog-info.yaml file.

This action is similar to the "Register existing component" flow in the Backstage UI, where you provide a URL to a catalog-info.yaml file describing catalog entities such as Components, Systems, Resources, APIs, Users, and Groups. The action will create a new Location entity that references the provided file; all entities defined within that file will be imported into the catalog.

A unique identifier (locationId) will be returned for the newly created Location. You can use this identifier in the future to unregister or delete the Location and all entities it owns.
`,
    schema: {
      input: (z) => z.object({
        locationUrl: z.string().describe(
          `URL reference to the catalog-info.yaml file that describes the entity to register. For example: https://github.com/backstage/demo/blob/master/catalog-info.yaml`
        )
      }),
      output: (z) => z.object({
        locationId: z.string().describe("The ID of the entity that was registered")
      })
    },
    action: async ({ input, credentials }) => {
      if (!isValidUrl(input.locationUrl)) {
        throw new errors.InputError(`${input.locationUrl} is an invalid URL`);
      }
      const result = await catalog.addLocation(
        {
          type: "url",
          target: input.locationUrl
        },
        {
          credentials
        }
      );
      return {
        output: {
          locationId: result.location.id
        }
      };
    }
  });
};

exports.createRegisterCatalogEntitiesAction = createRegisterCatalogEntitiesAction;
