// @backstage/plugin-notifications/dist/components/NotificationsFilters/NotificationsFilters.esm.js
import { jsx, Fragment, jsxs } from 'react/jsx-runtime';
import FormControl from '@material-ui/core/FormControl';
import Divider from '@material-ui/core/Divider';
import Grid from '@material-ui/core/Grid';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import Typography from '@material-ui/core/Typography';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { notificationsTranslationRef } from '../../translation.esm.js';

const ALL = "___all___";
const CreatedAfterOptions = {
  last24h: {
    labelKey: "filters.createdAfter.last24h",
    getDate: () => new Date(Date.now() - 24 * 3600 * 1e3)
  },
  lastWeek: {
    labelKey: "filters.createdAfter.lastWeek",
    getDate: () => new Date(Date.now() - 7 * 24 * 3600 * 1e3)
  },
  all: {
    labelKey: "filters.createdAfter.anyTime",
    getDate: () => /* @__PURE__ */ new Date(0)
  }
};
const SortByOptions = {
  newest: {
    labelKey: "filters.sortBy.newest",
    sortBy: {
      sort: "created",
      sortOrder: "desc"
    }
  },
  oldest: {
    labelKey: "filters.sortBy.oldest",
    sortBy: {
      sort: "created",
      sortOrder: "asc"
    }
  },
  topic: {
    labelKey: "filters.sortBy.topic",
    sortBy: {
      sort: "topic",
      sortOrder: "asc"
    }
  },
  origin: {
    labelKey: "filters.sortBy.origin",
    sortBy: {
      sort: "origin",
      sortOrder: "asc"
    }
  }
};
const getSortByText = (sortBy) => {
  if (sortBy?.sort === "created" && sortBy?.sortOrder === "asc") {
    return "oldest";
  }
  if (sortBy?.sort === "topic") {
    return "topic";
  }
  if (sortBy?.sort === "origin") {
    return "origin";
  }
  return "newest";
};
const NotificationsFilters = ({
  sorting,
  onSortingChanged,
  unreadOnly,
  onUnreadOnlyChanged,
  createdAfter,
  onCreatedAfterChanged,
  saved,
  onSavedChanged,
  severity,
  onSeverityChanged,
  topic,
  onTopicChanged,
  allTopics
}) => {
  const { t } = useTranslationRef(notificationsTranslationRef);
  const sortByText = getSortByText(sorting);
  const handleOnCreatedAfterChanged = (event) => {
    onCreatedAfterChanged(event.target.value);
  };
  const handleOnViewChanged = (event) => {
    if (event.target.value === "unread") {
      onUnreadOnlyChanged(true);
      onSavedChanged(void 0);
    } else if (event.target.value === "read") {
      onUnreadOnlyChanged(false);
      onSavedChanged(void 0);
    } else if (event.target.value === "saved") {
      onUnreadOnlyChanged(void 0);
      onSavedChanged(true);
    } else {
      onUnreadOnlyChanged(void 0);
      onSavedChanged(void 0);
    }
  };
  const handleOnSortByChanged = (event) => {
    const idx = event.target.value || "newest";
    const option = SortByOptions[idx];
    onSortingChanged({ ...option.sortBy });
  };
  let viewValue = "all";
  if (saved) {
    viewValue = "saved";
  } else if (unreadOnly) {
    viewValue = "unread";
  } else if (unreadOnly === false) {
    viewValue = "read";
  }
  const handleOnSeverityChanged = (event) => {
    const value = event.target.value || "normal";
    onSeverityChanged(value);
  };
  const handleOnTopicChanged = (event) => {
    const value = event.target.value;
    onTopicChanged(value === ALL ? void 0 : value);
  };
  const sortedAllTopics = (allTopics || []).sort((a, b) => a.localeCompare(b));
  return /* @__PURE__ */ jsx(Fragment, { children: /* @__PURE__ */ jsxs(Grid, { container: true, children: [
    /* @__PURE__ */ jsxs(Grid, { item: true, xs: 12, children: [
      /* @__PURE__ */ jsx(Typography, { variant: "h6", children: t("filters.title") }),
      /* @__PURE__ */ jsx(Divider, { variant: "fullWidth" })
    ] }),
    /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, children: /* @__PURE__ */ jsxs(FormControl, { fullWidth: true, variant: "outlined", size: "small", children: [
      /* @__PURE__ */ jsx(InputLabel, { id: "notifications-filter-view", children: t("filters.view.label") }),
      /* @__PURE__ */ jsxs(
        Select,
        {
          labelId: "notifications-filter-view",
          label: t("filters.view.label"),
          value: viewValue,
          onChange: handleOnViewChanged,
          children: [
            /* @__PURE__ */ jsx(MenuItem, { value: "unread", children: t("filters.view.unread") }),
            /* @__PURE__ */ jsx(MenuItem, { value: "read", children: t("filters.view.read") }),
            /* @__PURE__ */ jsx(MenuItem, { value: "saved", children: t("filters.view.saved") }),
            /* @__PURE__ */ jsx(MenuItem, { value: "all", children: t("filters.view.all") })
          ]
        }
      )
    ] }) }),
    /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, children: /* @__PURE__ */ jsxs(FormControl, { fullWidth: true, variant: "outlined", size: "small", children: [
      /* @__PURE__ */ jsx(InputLabel, { id: "notifications-filter-created", children: t("filters.createdAfter.label") }),
      /* @__PURE__ */ jsx(
        Select,
        {
          label: t("filters.createdAfter.label"),
          labelId: "notifications-filter-created",
          placeholder: t("filters.createdAfter.placeholder"),
          value: createdAfter,
          onChange: handleOnCreatedAfterChanged,
          children: Object.keys(CreatedAfterOptions).map((key) => /* @__PURE__ */ jsx(MenuItem, { value: key, children: t(
            CreatedAfterOptions[key].labelKey
          ) }, key))
        }
      )
    ] }) }),
    /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, children: /* @__PURE__ */ jsxs(FormControl, { fullWidth: true, variant: "outlined", size: "small", children: [
      /* @__PURE__ */ jsx(InputLabel, { id: "notifications-filter-sort", children: t("filters.sortBy.label") }),
      /* @__PURE__ */ jsx(
        Select,
        {
          label: t("filters.sortBy.label"),
          labelId: "notifications-filter-sort",
          placeholder: t("filters.sortBy.placeholder"),
          value: sortByText,
          onChange: handleOnSortByChanged,
          children: Object.keys(SortByOptions).map((key) => /* @__PURE__ */ jsx(MenuItem, { value: key, children: t(SortByOptions[key].labelKey) }, key))
        }
      )
    ] }) }),
    /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, children: /* @__PURE__ */ jsxs(FormControl, { fullWidth: true, variant: "outlined", size: "small", children: [
      /* @__PURE__ */ jsx(InputLabel, { id: "notifications-filter-severity", children: t("filters.severity.label") }),
      /* @__PURE__ */ jsx(
        Select,
        {
          label: t("filters.severity.label"),
          labelId: "notifications-filter-severity",
          value: severity,
          onChange: handleOnSeverityChanged,
          children: ["critical", "high", "normal", "low"].map((key) => /* @__PURE__ */ jsx(MenuItem, { value: key, children: t(`filters.severity.${key}`) }, key))
        }
      )
    ] }) }),
    /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, children: /* @__PURE__ */ jsxs(FormControl, { fullWidth: true, variant: "outlined", size: "small", children: [
      /* @__PURE__ */ jsx(InputLabel, { id: "notifications-filter-topic", children: t("filters.topic.label") }),
      /* @__PURE__ */ jsxs(
        Select,
        {
          label: t("filters.topic.label"),
          labelId: "notifications-filter-topic",
          value: topic ?? ALL,
          onChange: handleOnTopicChanged,
          children: [
            /* @__PURE__ */ jsx(MenuItem, { value: ALL, children: t("filters.topic.anyTopic") }, ALL),
            sortedAllTopics.map((item) => /* @__PURE__ */ jsx(MenuItem, { value: item, children: item }, item))
          ]
        }
      )
    ] }) })
  ] }) });
};

export { CreatedAfterOptions, NotificationsFilters, SortByOptions };
