// @backstage/cli/dist/modules/info/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var yargs = require('yargs');
var factory = require('../../wiring/factory.cjs.js');
var lazy = require('../../lib/lazy.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var yargs__default = /*#__PURE__*/_interopDefaultCompat(yargs);

var index = factory.createCliPlugin({
  pluginId: "info",
  init: async (reg) => {
    reg.addCommand({
      path: ["info"],
      description: "Show helpful information for debugging and reporting bugs",
      execute: async ({ args }) => {
        const argv = await yargs__default.default().options({
          include: {
            type: "string",
            array: true,
            default: [],
            description: "Glob patterns for additional packages to include (e.g., @spotify/backstage*)"
          },
          format: {
            type: "string",
            choices: ["text", "json"],
            default: "text",
            description: "Output format (text or json)"
          }
        }).help().parse(args);
        await lazy.lazy(() => import('./commands/info.cjs.js'), "default")(argv);
      }
    });
  }
});

exports.default = index;
