// @backstage/backend-defaults/dist/alpha/entrypoints/instanceMetadata/instanceMetadataServiceFactory.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');

const instanceMetadataServiceRef = backendPluginApi.createServiceRef({
  id: "core.instanceMetadata"
});
const instanceMetadataServiceFactory = backendPluginApi.createServiceFactory({
  service: instanceMetadataServiceRef,
  deps: {
    instanceMetadata: backendPluginApi.coreServices.rootInstanceMetadata
  },
  factory: async ({ instanceMetadata }) => {
    const plugins = await instanceMetadata.getInstalledPlugins();
    const features = [];
    for (const plugin of plugins) {
      features.push({
        type: "plugin",
        pluginId: plugin.pluginId
      });
      for (const module of plugin.modules) {
        features.push({
          type: "module",
          pluginId: plugin.pluginId,
          moduleId: module.moduleId
        });
      }
    }
    const service = {
      getInstalledFeatures: () => features
    };
    return service;
  }
});

exports.instanceMetadataServiceFactory = instanceMetadataServiceFactory;
exports.instanceMetadataServiceRef = instanceMetadataServiceRef;
