// @backstage/plugin-notifications/dist/components/NotificationsTable/SelectAll.esm.js
import { jsx } from 'react/jsx-runtime';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { makeStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';

const useStyles = makeStyles({
  label: {
    marginLeft: "0px",
    maxWidth: "2rem",
    "& span": {
      paddingRight: "0px",
      marginRight: "2px"
    }
  }
});
const SelectAll = ({
  count,
  totalCount,
  onSelectAll
}) => {
  const classes = useStyles();
  return /* @__PURE__ */ jsx(
    FormControlLabel,
    {
      label: count > 0 ? `(${count})` : void 0,
      className: classes.label,
      control: /* @__PURE__ */ jsx(Tooltip, { title: "Select all", children: /* @__PURE__ */ jsx(
        Checkbox,
        {
          color: "primary",
          disabled: !totalCount,
          checked: count > 0,
          indeterminate: count > 0 && totalCount !== count,
          onChange: onSelectAll
        }
      ) })
    }
  );
};

export { SelectAll };
