// @backstage/ui/dist/components/Grid/Grid.esm.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import clsx from 'clsx';
import { useStyles } from '../../hooks/useStyles.esm.js';
import { GridDefinition, GridItemDefinition } from './definition.esm.js';
import styles from './Grid.module.css.esm.js';
import { useSurface, SurfaceProvider } from '../../hooks/useSurface.esm.js';

const GridRoot = forwardRef((props, ref) => {
  const { surface: resolvedSurface } = useSurface({
    surface: props.surface
  });
  const { classNames, dataAttributes, utilityClasses, style, cleanedProps } = useStyles(GridDefinition, {
    columns: "auto",
    gap: "4",
    ...props,
    surface: resolvedSurface
    // Use resolved surface for data attribute
  });
  const { className, surface, ...rest } = cleanedProps;
  const content = /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      className: clsx(
        classNames.root,
        utilityClasses,
        styles[classNames.root],
        className
      ),
      style,
      ...dataAttributes,
      ...rest
    }
  );
  return resolvedSurface ? /* @__PURE__ */ jsx(SurfaceProvider, { surface: resolvedSurface, children: content }) : content;
});
const GridItem = forwardRef((props, ref) => {
  const { surface: resolvedSurface } = useSurface({
    surface: props.surface
  });
  const { classNames, dataAttributes, utilityClasses, style, cleanedProps } = useStyles(GridItemDefinition, {
    ...props,
    surface: resolvedSurface
    // Use resolved surface for data attribute
  });
  const { className, surface, ...rest } = cleanedProps;
  const content = /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      className: clsx(
        classNames.root,
        utilityClasses,
        styles[classNames.root],
        className
      ),
      style,
      ...dataAttributes,
      ...rest
    }
  );
  return resolvedSurface ? /* @__PURE__ */ jsx(SurfaceProvider, { surface: resolvedSurface, children: content }) : content;
});
const Grid = {
  Root: GridRoot,
  Item: GridItem
};

export { Grid };
