// @backstage/plugin-kubernetes-react/dist/components/ErrorReporting/ErrorReporting.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { useMemo } from 'react';
import { Table } from '@backstage/core-components';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { kubernetesReactTranslationRef } from '../../translation.esm.js';

const sortBySeverity = (a, b) => {
  if (a.error.severity < b.error.severity) {
    return 1;
  } else if (b.error.severity < a.error.severity) {
    return -1;
  }
  return 0;
};
const ErrorReporting = ({
  detectedErrors,
  clusters
}) => {
  const { t } = useTranslationRef(kubernetesReactTranslationRef);
  const errors = Array.from(detectedErrors.entries()).flatMap(([clusterName, resourceErrors]) => {
    return resourceErrors.map((e) => ({
      cluster: clusters.find((c) => c.name === clusterName),
      error: e
    }));
  }).sort(sortBySeverity);
  const columns = useMemo(
    () => [
      {
        title: String(t("errorReporting.columns.cluster")),
        width: "10%",
        render: (row) => row.cluster.title || row.cluster.name
      },
      {
        title: String(t("errorReporting.columns.namespace")),
        width: "10%",
        render: (row) => row.error.sourceRef.namespace
      },
      {
        title: String(t("errorReporting.columns.kind")),
        width: "10%",
        render: (row) => row.error.sourceRef.kind
      },
      {
        title: String(t("errorReporting.columns.name")),
        width: "30%",
        render: (row) => {
          return /* @__PURE__ */ jsxs(Fragment, { children: [
            row.error.sourceRef.name,
            " "
          ] });
        }
      },
      {
        title: String(t("errorReporting.columns.messages")),
        width: "40%",
        render: (row) => row.error.message
      }
    ],
    [t]
  );
  return /* @__PURE__ */ jsx(Fragment, { children: errors.length !== 0 && /* @__PURE__ */ jsx(
    Table,
    {
      title: t("errorReporting.title"),
      data: errors,
      columns,
      options: { paging: true, search: false, emptyRowsWhenPaging: false }
    }
  ) });
};

export { ErrorReporting };
