// @backstage/ui/dist/hooks/useSurface.esm.js
import { jsx } from 'react/jsx-runtime';
import { useContext } from 'react';
import { createVersionedContext, createVersionedValueMap } from '@backstage/version-bridge';

const SurfaceContext = createVersionedContext("surface-context");
function incrementSurface(surface) {
  if (!surface) return "0";
  if (surface === "0") return "1";
  if (surface === "1") return "2";
  if (surface === "2" || surface === "3") return "3";
  if (surface === "danger") return "danger";
  if (surface === "warning") return "warning";
  if (surface === "success") return "success";
  if (surface === "auto") return "1";
  return surface;
}
function resolveSurfaceForProvider(contextSurface, requestedSurface) {
  if (!requestedSurface) {
    return contextSurface;
  }
  if (typeof requestedSurface === "object") {
    return requestedSurface;
  }
  if (typeof contextSurface === "object") {
    if (requestedSurface === "auto") {
      return "0";
    }
    return requestedSurface;
  }
  if (requestedSurface === "auto") {
    return incrementSurface(contextSurface);
  }
  return requestedSurface;
}
function resolveSurfaceForConsumer(contextSurface, requestedSurface) {
  if (!requestedSurface) {
    return contextSurface;
  }
  if (typeof requestedSurface === "object") {
    return requestedSurface;
  }
  if (requestedSurface === "auto") {
    if (typeof contextSurface === "object") {
      return "0";
    }
    return contextSurface;
  }
  return requestedSurface;
}
const SurfaceProvider = ({
  surface,
  children
}) => {
  return /* @__PURE__ */ jsx(
    SurfaceContext.Provider,
    {
      value: createVersionedValueMap({ 1: { surface } }),
      children
    }
  );
};
const useSurface = (options) => {
  const value = useContext(SurfaceContext)?.atVersion(1);
  const context = value ?? { surface: void 0 };
  const isProvider = options?.surface !== void 0;
  const requestedSurface = options?.surface ?? options?.onSurface;
  const resolvedSurface = isProvider ? resolveSurfaceForProvider(context.surface, requestedSurface) : resolveSurfaceForConsumer(context.surface, requestedSurface);
  return {
    surface: resolvedSurface
  };
};

export { SurfaceProvider, resolveSurfaceForProvider, useSurface };
