// @backstage/plugin-kubernetes-react/dist/components/ErrorPanel/ErrorPanel.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import Typography from '@material-ui/core/Typography';
import { WarningPanel } from '@backstage/core-components';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { kubernetesReactTranslationRef } from '../../translation.esm.js';

const clustersWithErrorsToErrorMessage = (clustersWithErrors, t) => {
  return clustersWithErrors.map((c, i) => {
    return /* @__PURE__ */ jsxs("div", { children: [
      /* @__PURE__ */ jsx(Typography, { variant: "body2", children: t("errorPanel.clusterLabelValue", {
        cluster: c.cluster.title || c.cluster.name
      }) }),
      c.errors.map((e, j) => {
        return /* @__PURE__ */ jsx(Typography, { variant: "body2", children: e.errorType === "FETCH_ERROR" ? t("errorPanel.fetchError", {
          errorType: e.errorType,
          message: e.message
        }) : t("errorPanel.resourceError", {
          resourcePath: e.resourcePath ?? "",
          errorType: e.errorType,
          statusCode: String(e.statusCode ?? "")
        }) }, j);
      }),
      /* @__PURE__ */ jsx("br", {})
    ] }, i);
  });
};
const ErrorPanel = ({
  entityName,
  errorMessage,
  clustersWithErrors
}) => {
  const { t } = useTranslationRef(kubernetesReactTranslationRef);
  return /* @__PURE__ */ jsxs(
    WarningPanel,
    {
      title: t("errorPanel.title"),
      message: t("errorPanel.message", { entityName }),
      children: [
        clustersWithErrors && /* @__PURE__ */ jsxs("div", { children: [
          t("errorPanel.errorsLabel"),
          ":",
          " ",
          clustersWithErrorsToErrorMessage(clustersWithErrors, t)
        ] }),
        errorMessage && /* @__PURE__ */ jsxs(Typography, { variant: "body2", children: [
          t("errorPanel.errorsLabel"),
          ": ",
          errorMessage
        ] })
      ]
    }
  );
};

export { ErrorPanel };
