// @backstage/plugin-kubernetes-react/dist/components/Pods/FixDialog/FixDialog.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { useState } from 'react';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import IconButton from '@material-ui/core/IconButton';
import { makeStyles, createStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import CloseIcon from '@material-ui/icons/Close';
import HelpIcon from '@material-ui/icons/Help';
import OpenInNewIcon from '@material-ui/icons/OpenInNew';
import { PodLogs } from '../PodLogs/PodLogs.esm.js';
import '@material-ui/icons/Subject';
import '../../KubernetesDialog/KubernetesDialog.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { kubernetesReactTranslationRef } from '../../../translation.esm.js';
import 'react-use/esm/useAsync';
import '@backstage/core-plugin-api';
import '../../../api/types.esm.js';
import { Events } from '../Events/Events.esm.js';
import { LinkButton } from '@backstage/core-components';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';

const useStyles = makeStyles(
  (theme) => createStyles({
    closeButton: {
      position: "absolute",
      right: theme.spacing(1),
      top: theme.spacing(1),
      color: theme.palette.grey[500]
    }
  })
);
const FixDialog = ({
  open,
  pod,
  error,
  clusterName
}) => {
  const [isOpen, setOpen] = useState(!!open);
  const classes = useStyles();
  const { t } = useTranslationRef(kubernetesReactTranslationRef);
  const openDialog = () => {
    setOpen(true);
  };
  const closeDialog = () => {
    setOpen(false);
  };
  const pf = error.proposedFix;
  const dialogContent = () => {
    return /* @__PURE__ */ jsxs(Grid, { container: true, children: [
      /* @__PURE__ */ jsxs(Grid, { item: true, xs: 12, children: [
        /* @__PURE__ */ jsx(Typography, { variant: "h6", children: t("fixDialog.detectedError") }),
        /* @__PURE__ */ jsx(Typography, { children: error.message })
      ] }),
      /* @__PURE__ */ jsxs(Grid, { item: true, xs: 12, children: [
        /* @__PURE__ */ jsx(Typography, { variant: "h6", children: t("fixDialog.causeExplanation") }),
        /* @__PURE__ */ jsx(Typography, { children: error.proposedFix?.rootCauseExplanation ?? t("podsTable.unknown") })
      ] }),
      /* @__PURE__ */ jsxs(Grid, { item: true, xs: 12, children: [
        /* @__PURE__ */ jsx(Typography, { variant: "h6", children: t("fixDialog.fix") }),
        /* @__PURE__ */ jsx(List, { children: (error.proposedFix?.actions ?? []).map((fix, i) => {
          return /* @__PURE__ */ jsx(
            ListItem,
            {
              children: fix
            },
            `${pod.metadata?.name ?? t("podsTable.unknown")}-pf-${i}`
          );
        }) })
      ] }),
      pf && pf.type === "logs" && /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, children: /* @__PURE__ */ jsx(Typography, { variant: "h6", children: t("fixDialog.crashLogs") }) }),
        /* @__PURE__ */ jsx(Grid, { item: true, xs: 9, children: /* @__PURE__ */ jsx(
          PodLogs,
          {
            previous: true,
            containerScope: {
              podName: pod.metadata?.name ?? t("podsTable.unknown"),
              podNamespace: pod.metadata?.namespace ?? t("podsTable.unknown"),
              cluster: { name: clusterName },
              containerName: pf.container
            }
          }
        ) })
      ] }),
      pf && pf.type === "events" && /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(Grid, { item: true, xs: 12, children: /* @__PURE__ */ jsx(Typography, { variant: "h6", children: t("fixDialog.events") }) }),
        /* @__PURE__ */ jsx(Grid, { item: true, xs: 9, children: /* @__PURE__ */ jsx(
          Events,
          {
            warningEventsOnly: true,
            involvedObjectName: pod.metadata?.name ?? "",
            namespace: pod.metadata?.namespace ?? "",
            clusterName
          }
        ) })
      ] })
    ] });
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      Button,
      {
        variant: "outlined",
        "aria-label": t("fixDialog.ariaLabels.fixIssue"),
        component: "label",
        onClick: openDialog,
        startIcon: /* @__PURE__ */ jsx(HelpIcon, {}),
        children: t("fixDialog.helpButton")
      }
    ),
    /* @__PURE__ */ jsxs(Dialog, { maxWidth: "xl", fullWidth: true, open: isOpen, onClose: closeDialog, children: [
      /* @__PURE__ */ jsxs(DialogTitle, { id: "dialog-title", children: [
        t("fixDialog.title", {
          podName: pod.metadata?.name ?? "",
          errorType: error.type
        }),
        /* @__PURE__ */ jsx(
          IconButton,
          {
            "aria-label": t("fixDialog.ariaLabels.close"),
            className: classes.closeButton,
            onClick: closeDialog,
            children: /* @__PURE__ */ jsx(CloseIcon, {})
          }
        )
      ] }),
      /* @__PURE__ */ jsx(DialogContent, { children: dialogContent() }),
      /* @__PURE__ */ jsx(DialogActions, { children: pf && pf.type === "docs" && /* @__PURE__ */ jsx(
        LinkButton,
        {
          to: pf.docsLink,
          variant: "outlined",
          startIcon: /* @__PURE__ */ jsx(OpenInNewIcon, {}),
          target: "_blank",
          rel: "noopener",
          children: t("fixDialog.openDocs")
        }
      ) })
    ] })
  ] });
};

export { FixDialog };
